### PR TITLE
Don't reload routes twice on admin code changes.

### PR DESCRIPTION
**Pull request by kaspth**
If someone makes a change in an `app/admin` file, Rails will trigger a
routes reload by eval'ing `routes.rb` which in turn will call `routes`
on the Active Admin application.

Later, because we've inserted ourselves into the reloading through the
`to_prepare` and `to_complete` callbacks, `to_prepare` will unload our
namespaces while `to_complete` will retrigger a routes and application
reload.

Thus we can reliably skip the first routes reload whenver the reloader
is `updated?`. Because `execute_if_updated` is called if it is updated
the routes will be reloaded at that time. When `execute_if_updated` is
reloading the routes the reloader is no longer `updated?`. Finally, if
the code change wasn't in `app/admin` our reloader won't be `updated?`
there either.

### CODE REVIEW

File: lib/active_admin/application.rb [hunk 1]
```diff
@@ -250,6 +256,8 @@ def controllers_for_filters
       controllers
     end
 
+    attr_writer :routes_reloader # :nodoc:
```
Comment (kaspth) [inline lib/active_admin/application.rb]: Don't intend this to a public API, but I'm unsure how you keep methods in the public namespace from showing up in documentation.

Comment (kaspth) [inline lib/active_admin/application.rb]: Alternatively we could return the capture the reloader with `@routes_reloader = attach_reloader` in `prepare!`. That will just force a local variable capture into `attach_reloader`. Let me know which you prefer 😁

**Comment by timoschilling** [comment]
can some rebase this to get the tests green

**Comment by kaspth** [comment]
@timoschilling rebased 👍 — I'll keep an eye on the build ❤️

File: lib/active_admin/application.rb [hunk 2]
```diff
@@ -233,7 +233,13 @@ def router
     # One-liner called by user's config/routes.rb file
     def routes(rails_router)
       load!
-      router.apply(rails_router)
+
+      # If our reloader has been updated, we now that it will trigger
```
Comment (varyonic) [inline lib/active_admin/application.rb]: know

Comment (kaspth) [inline lib/active_admin/application.rb]: Knowing what I now know, I'd have spelled this differently 🤓😄

**Comment by kaspth** [comment]
Not sure why exactly this change causes the build to fail, if you have pointers I'll gladly take them 👍 

I'll try to investigate soon.

**Comment by deivid-rodriguez** [comment]
@kaspth I'm working on #4805 and I just run into this same failure. Once #4805 lands, rebasing this PR should get it green.

**Comment by deivid-rodriguez** [comment]
For the record, the fix for this particular failure should be https://github.com/activeadmin/activeadmin/pull/4805/commits/69ebfe66757c2c18e7b68e51c0333039b4ecb778.

**Comment by kaspth** [comment]
@deivid-rodriguez I love you 😄

**Comment by deivid-rodriguez** [comment]
@kaspth https://github.com/rails/rails/pull/28038 I love you too :heart_eyes:

**Comment by deivid-rodriguez** [comment]
@kaspth Could you rebase this now?

**Comment by kaspth** [comment]
@deivid-rodriguez rebased! Thanks for fixing the order dependence in the tests 😊

**Comment by kaspth** [comment]
Well, some of the builds turned green 😄

**Comment by Fivell** [comment]
@deivid-rodriguez , there so much failed specs, that looks not like random issue

**Comment by deivid-rodriguez** [comment]
I wanted to have a look but didn't get the chance. But yes... although some times one failures triggers a lot of the failures...

**Comment by Fivell** [comment]
@deivid-rodriguez , let's rerun build and see if something changes

**Comment by deivid-rodriguez** [comment]
@kaspth Fixed one more issue affecting this in #4852. Could you try rebasing one more time?

**Comment by kaspth** [comment]
@deivid-rodriguez you rock, thanks so much! I've rebased again 🤘

**Comment by kaspth** [comment]
All green! 💸💸💸

**Comment by Fivell** [comment]
@kaspth , any idea how to add spec for this to not break it in future ?

**Comment by kaspth** [comment]
@Fivell not entirely sure, but here's a high-level quite-handwavy version:

Generate a template app to keep in the test folder, copy that to a tmp folder, cd into it, overwrite `@routes_reloader` with one that supers but also tracks the number of times we've triggered reloading, then boot a rails server, issue a request to the server, overwrite an admin file, issue the request again, then assert that the mocked `@routes_reloader`'s `reload` was only called once.

**Comment by Fivell** [comment]
@kaspth, that will be too much I think :)

**Comment by Fivell** [comment]
@varyonic , let's give this a try ?

**Comment by deivid-rodriguez** [comment]
We have a `class-reloading` cucumber profile. Maybe it has the right setup to test this? Have someone tested this manually anyways? I mean other than @kaspth of course :)

**Comment by javierjulio** [comment]
Sounds like a lot but it could be doable. If what @deivid-rodriguez suggests is right then worth a try. Otherwise I've rummaged through a lot of Rails to trust @kaspth's work. His input here is very valuable so I'm fine with taking it as is but I understand the concern about wanting a test.

**Comment by kaspth** [comment]
I haven't used cucumber before, so I'm afraid trying my hand at a full stack test is a bit much for me at the moment. Though I've just enabled PR edits from maintainers so if you want to give it a go you can! 😊

We've also just upgraded our app to Active Admin 1.0 — works wonderfully, thanks! 💪

**Comment by deivid-rodriguez** [comment]
I'm fine with taking this as is as well! But our cucumbers seem to have the right setup to test this, so I wanted to at least give it a try.

**Comment by deivid-rodriguez** [comment]
@kaspth I had a look at this, but can't get it to work as I expect. This is what I'm doing:

* Generate a sample application and edit `config/routes.rb` dropping `puts "Reloading routes"` on top of it.
* Boot it and log into the admin.
* Edit a file in `app/admin`.
* Reload the page.

I seem to always get the message printed twice both with and without your changes. Any ideas?

**Comment by kaspth** [comment]
I’m no longer with the project that required this change so I’ll just close this. Thanks for all the help and reviews and sorry for not getting back to you! ❤️

**Pull request closed without merging**